packages/compiler/src/core/checker.ts

```diff
@@ -5165,6 +5165,18 @@ export function createChecker(program: Program, resolver: NameResolver): Checker
           target: node.targetType,
         }),
       );
+    } else if (
```

Comment:
If someone wants to take over this PR during my leave I think the remaining items are:
- this probably should also error for union expression
- add tests

Reply:
@allenjzhang